File: app/src/main/java/com/example/studytime/data/repository/StudyHallRepository.kt
package com.example.studytime.data.repository

import com.example.studytime.data.model.StudyHall
import com.example.studytime.utils.Resource
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.tasks.await
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class StudyHallRepository @Inject constructor(private val db: FirebaseFirestore) {

    suspend fun getHalls(): Resource<List<StudyHall>> = try {
        val snapshot = db.collection("studyHalls").get().await()
        val halls = snapshot.documents.mapNotNull { doc ->
            doc.toObject(StudyHall::class.java)?.copy(id = doc.id)
        }
        Resource.Success(halls)
    } catch (e: Exception) {
        Resource.Error(e.message ?: "Could not load study halls")
    }

    suspend fun getBookedSlots(hallId: String, tableNumber: Int, seatNumber: Int, date: String): Resource<Set<Int>> = try {
        val snapshot = db.collection("bookings")
            .whereEqualTo("hallId", hallId)
            .whereEqualTo("tableNumber", tableNumber)
            .whereEqualTo("seatNumber", seatNumber)
            .whereEqualTo("date", date)
            .whereEqualTo("status", "active")
            .get().await()
        val booked = snapshot.documents.mapNotNull {
            it.getLong("startHour")?.toInt()
        }.toSet()
        Resource.Success(booked)
    } catch (e: Exception) {
        Resource.Error(e.message ?: "Could not load time slots")
    }
}

File: lib/data/repositories/study_hall_repository.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';
import '../models/study_hall.dart';

part 'study_hall_repository.g.dart';

@Riverpod(keepAlive: true)
StudyHallRepository studyHallRepository(Ref ref) =>
    StudyHallRepository(FirebaseFirestore.instance);

class StudyHallRepository {
  final FirebaseFirestore _db;
  const StudyHallRepository(this._db);

  Future<List<StudyHall>> getHalls() async {
    final snap = await _db.collection('studyHalls').get();
    return snap.docs
        .map((d) => StudyHall.fromJson(d.data(), id: d.id))
        .toList();
  }

  Future<Set<int>> getBookedSlots({
    required String hallId,
    required int tableNumber,
    required int seatNumber,
    required String date,
  }) async {
    final snap = await _db
        .collection('bookings')
        .where('hallId', isEqualTo: hallId)
        .where('tableNumber', isEqualTo: tableNumber)
        .where('seatNumber', isEqualTo: seatNumber)
        .where('date', isEqualTo: date)
        .where('status', isEqualTo: 'active')
        .get();
    return snap.docs
        .map((d) => (d.data()['startHour'] as num).toInt())
        .toSet();
  }
}
